### ISSUE TITLE
old-time fails to detect the c compiler on Windows

### ISSUE THREAD

**Issue by arsdragonfly** [haskell/old-time#5]
When trying to install stylish-haskell and ghc-mod, an error occurs saying that old-time cannot be installed. Running "cabal install old-time" yields the following result:

````
λ cabal install old-time
Resolving dependencies...
Configuring old-time-1.1.0.3...
Failed to install old-time-1.1.0.3
Build log ( C:\Users\arsdr\AppData\Roaming\cabal\logs\old-time-1.1.0.3.log ):
Configuring old-time-1.1.0.3...
bash.exe: warning: could not find /tmp, please create!
configure: WARNING: unrecognized options: --with-compiler
checking for gcc... G:\Haskell\802E01~1.1艔
checking whether the C compiler works... no
configure: error: in `/cygdrive/c/Users/arsdr/AppData/Local/Temp/cabal-tmp-13348/old-time-1.1.0.3':
configure: error: C compiler cannot create executables
See `config.log' for more details
cabal: Leaving directory 'C:\Users\arsdr\AppData\Local\Temp\cabal-tmp-13348\old-time-1.1.0.3'
cabal: Error: some packages failed to install:
old-time-1.1.0.3 failed during the configure step. The exception was:
ExitFailure 77
````

Note the line "checking for gcc... G:\Haskell\802E01~1.1艔": the path seems to be screwed up (the directory for my Haskell Platform installation is G:\Haskell\8.0.1\). Also, there seems to be a similar question here on StackOverflow: [here](http://stackoverflow.com/questions/40260323/installation-error-with-haskell-cabal#)
I'd suggest that there's something wrong in the configure script. Hopefully it may get fixed.

**Comment by madscientist42** [comment]
Still ongoing with latest version of Haskell and Cabal.  Seriously, guys...if you're doing things that need bash/sh is this portable???

**Comment by ciez** [comment]
````
Configuring old-time-1.1.0.3...
configure: WARNING: unrecognized options: --with-compiler
checking for gcc... C:\PROGRA~1\GHC-80~1.2\mingw\bin\gcc.exe
checking whether the C compiler works... no
````

ghc is installed in C:\Program Files\ghc-8.0.2

any hints re: where 
C:\PROGRA\~1\GHC-80\~1.2\mingw\bin\gcc.exe 
comes from?

### solved:
cabal get old-time-1.1.0.3

insert _ac_ct_CC=gcc_ in configure beginning @ 2258:

````
if ${ac_cv_prog_ac_ct_CC+:} false; then :
  $as_echo_n "(cached) " >&6
else
  ac_ct_CC=gcc           ## insert this line ##
  if test -n "$ac_ct_CC"; then
  ac_cv_prog_ac_ct_CC="$ac_ct_CC" # Let the user override the test.
````

then build from cygwin

**Comment by hvr** [comment]
@clez which cabal version are you using? It seems to be passing the deprecated `--with-compiler` flag to the `./configure` script

**Comment by ciez** [comment]
@hvr

cabal-install version 1.24.0.2
compiled using version 1.24.2.0 of the Cabal library

The Glorious Glasgow Haskell Compilation System, version 8.0.2

This is a temporary installation, old-time and ghc-mod are now installed. This is no longer an issue for me at the moment. Thank you for prompt reply.

**Comment by vanufryiuk** [comment]
The same issue under the same cabal version on Windows 10 x64, Haskell Platform Full x64 8.0.2.

**Comment by hvr** [comment]
Regarding the question where the `CC=` value comes from that is passed to `./configure`; that's supposed to be the value that `ghc --info` provides as compiler command, e.g.

```
 ,("C compiler command","/usr/bin/gcc")
```

@vanufryiuk can you lookup what `ghc --info` would tell you about GCC?

**Comment by Mistuke** [comment]
I think there might be two separate issues here.

On Windows, the compiler will by default use the bundled GCC. The settings file should have an entry saying `$TopDir\..\mingw\bin\gcc.exe` or similar. `$TopDir` is gotten at startup by getting the location of `ghc` and stripping the `ghc.exe` from it, on Vista+ we explicit ask for NT filepaths, so this api will never give us a DOS shortname. After this `configure` must called with `CC=..` and subsequently invokes shell commands with this path or Platform must be changing the PATH, and your path inheritance settings for cygwin from Windows would then include it. Which one it is I'm not sure but that doesn't matter.

Once the path hits the shell a path conversion is done, it seems your shell is configured to convert to dos shortnames. so you end up with `G:\Haskell\802E01~1.1艔` during this conversion. I'm not sure why, it could either be a bug in `cygpath` or we were leaving non-printable characters at the end of the string. This conversion is likely being done because the path contains spaces, which some tools would have a hard time with.

So I think this is the first issue.

Looking at your second log, `checking for gcc... C:\PROGRA~1\GHC-80~1.2\mingw\bin\gcc.exe`
the conversion seems to have been fixed. Now this issue you're describing looks pretty similar to the Creators Update issue. Since you're on Windows 10 this may well be the case. Do you have a patched Haskell Platform 8.0.2? You can easily check this by seeing if `ghci` runs.

If not, run `C:\PROGRA~1\GHC-80~1.2\mingw\bin\gcc.exe` outside of Cygwin and see what error you get. Cygwin swallows exceptions to report them in `$?` so it hides the actual windows error dialogs.

**Comment by Mistuke** [comment]
Also, replacing the bundled GCC with that of Cygwin means you're also switching a GCC producing native windows binaries to one producing native POSIX binaries. You'll have a hard time compiling packages such as `Win32` with this version.

**Comment by pold500** [comment]
Is there gonna be any pull request that fix that issue?
Why developers treat Windows users as a 3rd class citizens?

**Comment by Mistuke** [comment]
@pold500 Can you paste the full output of configure? Which platform are you using?

**Comment by pold500** [comment]
@Mistuke Platform is Windows 10. Full output:
 Configuring old-time-1.1.0.3...
configure: WARNING: unrecognized options: --with-compiler
checking for gcc... C:\PROGRA~1\HASKEL~1\826561~1.1\mingw\bin\gcc.exe
checking whether the C compiler works... no
configure: error: in `/cygdrive/c/Users/userx/old-time-1.1.0.3':
configure: error: C compiler cannot create executables
See `config.log' for more details

config.log: https://pastebin.com/wSRBBLGR

**Comment by Mistuke** [comment]
weird, why does your log identify itself as Cygwin, and you have the `/cygdrive` prefix for your automounts.

Are you mixing Cygwin and msys2? what command did you use? just simply `cabal install old-time`?

**Comment by Mistuke** [comment]
I've been trying to reproduce this with platform, but I have not yet been able to. Could you describe a bit how your environment looks like?

**Comment by pold500** [comment]
@Mistuke, I tried the proposed solutions from step 2 and looks like it's working. My environment is Win10 machine with cygwin and mingw.

**Comment by Mistuke** [comment]
The solution in step 2 is a dangerous one. Since it's now picking up your system gcc which is a cygwin one designed to make posix binaries.

You never answered how you are doing the compilation. I would recommend you use the environment set up in platform, which uses msys2 instead of mixing the two.

Is there a particular reason you're using cygwin over msys2? Be aware that mixing cygwin and msys2 applications can produce weirdness as msys2 is cygwin derived but not exactly cygwin.

**Comment by pold500** [comment]
@Mitsuke. First time, I just installed Haskell platform (core). I omitted step 2. Then I tried "cabal install hindent". It has failed with "compiler can't produce binaries" error. So I built binaries by myself, using cygwin's autoconf and Setup.exe compiled from Setup.hs. I had to manually edit configure file to be able to build old-time library.
Second time I had edit the cabal conf file and tried to reinstall the package after manually removing it with ghc-pkg. Looks like it's worked but that's not a clean experiment, of course.
